aten/src/ATen/native/cuda/BinaryAddSubKernel.cu

```diff
@@ -19,12 +20,43 @@ struct AddFunctor {
     accscalar_t alpha;
 };
 
+template<typename scalar_t, typename accscalar_t, int SCALAR_ARG>
+struct AddScalarFunctor {
+  static_assert(SCALAR_ARG == 1 || SCALAR_ARG == 2, "SCALAR_ARG must be either 1 or 2");
+  AddScalarFunctor(accscalar_t alpha, accscalar_t b): alpha(alpha), b(b) {}
+  __device__ __forceinline__ scalar_t operator() (const scalar_t a) const {
+    return static_cast<scalar_t>(SCALAR_ARG == 1 ? b + alpha * a : a + alpha * b);
+  }
+  private:
+    accscalar_t alpha;
+    accscalar_t b;
+};
+
 void add_kernel_cuda(TensorIteratorBase& iter, const Scalar& alpha_scalar) {
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kHalf, kBool, kBFloat16, iter.common_dtype(), "add_cuda/sub_cuda", [&]() {
-    using accscalar_t = at::acc_type<scalar_t, true>;
-    AddFunctor<scalar_t, accscalar_t> f(alpha_scalar.to<accscalar_t>());
-    gpu_kernel_with_scalars(iter, f);
-  });
+  if (!isIntegralType(iter.common_dtype(), /* includeBool */ true) && (iter.is_cpu_scalar(1) || iter.is_cpu_scalar(2))) {
+    // if common dtype is half the scalar constant can overflow in half precision, and yet the result can
+    // still be representable in the half dtype. Cast scalar to acc_type to have better accuracy.
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kHalf, kBool, kBFloat16, iter.common_dtype(), "add_cuda/sub_cuda", [&]() {
```

Comment:
iter.common_dtype cannot be integral here, right? So you don't need to dispatch to ALL types, only floating and complex and 3?

Reply:
Absolutely right, thanks.